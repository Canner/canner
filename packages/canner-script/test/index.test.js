/** @jsx builder */

// eslint-disable-next-line
import builder, {configure} from '../src'
import path from 'path';
import Sorter from './toolbar/sorter';
import Pagination from './toolbar/pagination';
import Filter from './toolbar/filter';

configure({
  visitorManager: {
    defaultVisitors: []
  }
});

describe('builder', () => {
  it('throw error', () => {
    expect(() => <myType />).toThrow(/unsupported type/);
  });

  describe('string', () => {
    it('should work', () => {
      expect(<string />).toMatchObject({
        type: 'string',
        ui: 'input',
        packageName: '@canner/antd-string-input'
      });
    });
  });

  describe('boolean', () => {
    it('should work', () => {
      expect(<boolean />).toMatchObject({
        type: 'boolean',
        ui: 'switch'
      });
    });
  });

  describe('number', () => {
    it('should work', () => {
      expect(<number />).toMatchObject({
        type: 'number',
        ui: 'input'
      });
    });
  });

  describe('relation', () => {
    it('should work', () => {
      expect(<relation relation={{
        type: 'toOne',
        to: 'users'
      }} />).toMatchObject({
        type: 'relation',
        ui: 'singleSelect',
        relation: {
          type: 'toOne',
          to: 'users'
        }
      });
    });

    it('should have toolbar attrs', () => {
      expect(<relation relation={{
        type: 'toOne',
        to: 'users'
      }}>
        <toolbar>
          <pagination />
        </toolbar>
      </relation>).toMatchObject({
        type: 'relation',
        ui: 'singleSelect',
        relation: {
          type: 'toOne',
          to: 'users'
        },
        toolbar: {
          pagination: {
            type: 'pagination'
          }
        }
      });
    });
  });

  describe('array', () => {
    it('array of string without keyName', () => {
      const array = (<array>
        <string />
      </array>);
      expect(array).toMatchObject({
        type: 'array',
        ui: 'table',
        items: {
          type: 'string'
        }
      });
    });

    it('array of string with keyName', () => {
      const array = (<array>
        <string keyName="test"/>
      </array>);
      expect(array).toMatchObject({
        type: 'array',
        ui: 'table',
        items: {
          type: 'object',
          items: {
            test: {
              type: 'string',
              keyName: 'test'
            }
          }
        }
      });
    });

    it('auto object in array', () => {
      const array = (<array>
        <string keyName="test" />
        <string keyName="test2" />
      </array>);
      expect(array).toMatchObject({
        type: 'array',
        ui: 'table',
        items: {
          type: 'object',
          items: {
            test: {
              type: 'string',
              keyName: 'test'
            },
            test2: {
              type: 'string',
              keyName: 'test2'
            }
          }
        }
      });
    });

    it('manual object without keyName in array', () => {
      const array = (<array>
        <object>
          <string keyName="test" />
        </object>
      </array>);
      expect(array).toMatchObject({
        type: 'array',
        ui: 'table',
        items: {
          type: 'object',
          items: {
            test: {
              type: 'string',
              keyName: 'test'
            }
          }
        }
      });
    });

    it('manual object with keyName in array', () => {
      const array = (<array>
        <object keyName="obj">
          <string keyName="test" />
        </object>
      </array>);
      expect(array).toMatchObject({
        type: 'array',
        ui: 'table',
        items: {
          type: 'object',
          items: {
            obj: {
              keyName: 'obj',
              type: 'object',
              items: {
                test: {
                  type: 'string',
                  keyName: 'test'
                }
              }
            }
          }
        }
      });
    });
  });

  describe('object', () => {
    it('should work', () => {
      const object = <object>
        <string keyName="title"/>
      </object>
      expect(object).toMatchObject({
        type: 'object',
        ui: 'fieldset',
        items: {
          title: {
            type: 'string',
            keyName: 'title'
          }
        }
      });
    });
  });

  describe('toolbar', () => {
    it('should work', () => {
      expect(<array>
        <toolbar>
          <sorter component={Sorter} />
          <pagination component={Pagination} />
          <filter component={Filter} />
        </toolbar>
      </array>).toMatchObject({
        type: 'array',
        toolbar: {
          sorter: {
            component: Sorter,
            type: 'sorter'
          },
          pagination: {
            component: Pagination,
            type: 'pagination'
          },
          filter: {
            component: Filter,
            type: 'filter'
          }
        }
      });
    });
  });

  describe('filter', () => {
    it('should work', () => {
      expect(
        <toolbar>
          <filter>
            <textFilter label="Title" field="title"/>
            <numberFilter label="Clicks" field="clicks"/>
            <dateFilter label="CreatedDate" field="createdDate"/>
            <relationSelectFilter label="Author" field="author" textCol="name"/>
            <selectFilter
              label="Status"
              options={[{
                  text: 'Draft',
                  condition: {
                    draft_eq: true,
                    trash_eq: false
                  }
              }, {
                text: 'All',
                condition: {}
              }]}
            />
          </filter>
        </toolbar>).toMatchObject({
            filter: {
              filters: [{
              type: 'text',
              label: 'Title',
              field: 'title' 
            }, {
              type: 'number',
              label: 'Clicks',
              field: 'clicks'
            }, {
              type: 'date',
              label: 'CreatedDate',
              field: 'createdDate'
            }, {
              type: 'relation',
              label: 'Author',
              field: 'author',
              textCol: 'name'
            }, {
              type: 'select',
              label: 'Status',
              options: [{
                text: 'Draft',
                condition: {
                  draft_eq: true,
                  trash_eq: false
                }
            }, {
              text: 'All',
              condition: {}
            }]
            }]}
        });
    });
  });

  describe('actions', () => {
    it('should work', () => {
      expect(<toolbar>
        <actions exportButton importButton filterButton />
      </toolbar>).toMatchObject({actions: {
        exportButton: true,
        importButton: true,
        filterButton: true
      }});
    });
  });

  describe('geoPoint', () => {
    it('should work', () => {
      expect(<geoPoint keyName="map"/>).toMatchObject({
        type: 'geoPoint',
        ui: 'default',
        keyName: 'map'
      });
    });
  });

  describe('dateTime', () => {
    it('should work', () => {
      expect(<dateTime keyName="date"/>).toMatchObject({
        type: 'dateTime',
        ui: 'dateTime',
        keyName: 'date'
      });
    });
  });

  describe('file', () => {
    it('should work', () => {
      expect(<file keyName="file"/>).toMatchObject({
        type: 'file',
        ui: 'file',
        keyName: 'file'
      });
    });
  });

  describe('image', () => {
    it('should work', () => {
      expect(<image keyName="image"/>).toMatchObject({
        type: 'image',
        ui: 'image',
        keyName: 'image'
      });
    });
  });

  describe('chart', () => {
    it('should work', () => {
      expect(<chart keyName="chart"/>).toMatchObject({
        type: 'chart',
        ui: 'line',
        keyName: 'chart'
      });
    });

    it('should work with children', () => {
      expect(
        <chart keyName="chart">
           <toolbar>
              <filter />
          </toolbar>
        </chart>
      ).toMatchObject({
        type: 'chart',
        ui: 'line',
        keyName: 'chart',
        toolbar: {
          filter: {
            "type": "filter"
          }
        }
      });
    });

    it('should have keyName', () => {
      expect(<chart />).toHaveProperty('keyName');
    });
  });

  describe('indicator', () => {
    it('should work', () => {
      expect(<indicator keyName="indicator"/>).toMatchObject({
        type: 'indicator',
        ui: 'amount',
        keyName: 'indicator'
      });
    });

    it('should work with children', () => {
      expect(
        <indicator keyName="amount">
           <toolbar>
              <filter />
          </toolbar>
        </indicator>
      ).toMatchObject({
        type: 'indicator',
        ui: 'amount',
        keyName: 'amount',
        toolbar: {
          filter: {
            "type": "filter"
          }
        }
      });
    });

    it('should have keyName', () => {
      expect(<indicator />).toHaveProperty('keyName');
    });
  });

  describe('enum', () => {
    it('should work', () => {
      expect(<enum keyName="enum" values={['A', 'B', 'C']}/>).toMatchObject({
        type: 'enum',
        ui: 'select',
        keyName: 'enum',
        values: ['A', 'B', 'C']
      });
    });
  });

  describe('root', () => {
    it('should work', () => {
      const root = <root dict={{en: {title: 'title'}}} storage="storage">
        <object keyName="info" >
          <string keyName="name"/>
        </object>
      </root>
      expect(root.storages).toMatchObject({
        info: 'storage'
      });
      expect(root.schema).toMatchObject({
        info: {
          type: 'object',
          ui: 'fieldset',
          items: {
            name: {
              keyName: 'name',
              ui: 'input',
              type: 'string'
            }
          }
        }
      });
      expect(root.dict).toMatchObject({en: {title: 'title'}});
    });

    it('should work with <page>child', () => {
      const root = <root>
        <page keyName="overview">
          <chart keyName="line" ui="line" />
        </page>
      </root>
      expect(root.pageSchema).toMatchObject({
        keyName: 'overview',
        type: 'page',
        items: {
            line: {
                type: 'chart',
                ui: 'line',
                keyName: 'line'
            }
        }
      });
    });
  });


  describe('page', () => {
    it('should work', () => {
      const page = <page keyName="overview">
        <indicator ui="amount" keyName="key1"/>
        <indicator ui="list" keyName="key2"/>
        <chart ui="bar" keyName="key3"/>
      </page>
      expect(page).toMatchObject({
        keyName: 'overview',
        type: 'page',
        items: {
          key1: {
              type: 'indicator',
              ui: 'amount',
              keyName: 'key1'
          },
          key2: {
              type: 'indicator',
              ui: 'list',
              keyName: 'key2'
          },
          key3: {
              type: 'chart',
              ui: 'bar',
              keyName: 'key3'
          }
        }
      });
    });
  });

  describe('children with map', () => {
    it('should work', () => {
      const schema = <root>
        {['info', 'info2'].map(name => <object key={name} keyName={name}>
          <string keyName={`${name}-string`}></string>
        </object>)}
      </root>;

      expect(schema.schema).toMatchObject({
        info: {
          type: 'object',
          items: {
            'info-string': {
              type: 'string',
              keyName: 'info-string'
            }
          }
        },
        info2: {
          type: 'object',
          items: {
            'info2-string': {
              type: 'string',
              keyName: 'info2-string'
            }
          }
        }
      });
    });
  });

  it('blog schema', () => {
    const blog = require('./blog');
    expect(blog).toMatchSnapshot();
  });

  it('function', () => {
    // eslint-disable-next-line
    const Variants = ({attributes, children}) => (
      <object {...attributes}>
        <string keyName="title" /> 
        {children}
      </object>
    )
    expect(<Variants keyName="variants">
      <string keyName="desc" />
    </Variants>).toMatchSnapshot();
  });

  describe('canner.def.js', () => {
    it('function builder', () => {
      const packageName = path.resolve(__dirname, 'fake-object-variants');
      // eslint-disable-next-line
      expect(<object keyName="variants" packageName={packageName} builder={require(`${packageName}/canner.def.js`)}>
        <string keyName="desc" />
      </object>).toMatchObject({
        "description": "",
        "items": {
          "a": {
            "keyName": "a",
            "packageName": "@canner/antd-string-input",
          },
          "d": {
            "keyName": "d",
            "packageName": "@canner/antd-string-input",
          },
          "desc": {
            "keyName": "desc",
            "packageName": "@canner/antd-string-input",
          },
        },
        "keyName": "variants",
        "packageName": packageName,
        "type": "object",
        "ui": "fieldset",
      });
    });

    it('require builder', () => {
      const packageName = path.resolve(__dirname, 'fake-object-variants');
      // eslint-disable-next-line
      expect(<object keyName="variants" packageName={packageName} builder={require(`${packageName}/canner.def.js`)}>
        <string keyName="desc" />
      </object>).toMatchObject({
        "items": {
          "a": {
            "description": "",
            "keyName": "a",
            "packageName": "@canner/antd-string-input",
            "title": "",
            "type": "string",
            "ui": "input",
          },
          "d": {
            "description": "",
            "keyName": "d",
            "packageName": "@canner/antd-string-input",
            "title": "",
            "type": "string",
            "ui": "input",
          },
          "desc": {
            "description": "",
            "keyName": "desc",
            "packageName": "@canner/antd-string-input",
            "title": "",
            "type": "string",
            "ui": "input",
            
          },
        },
        "keyName": "variants",
        "packageName": packageName,
        "title": "",
        "type": "object",
        "ui": "fieldset",
        
      });
    });
  })

  it('map chilren', () => {
    const list = ['a', 'b', 'c'];
    expect(<object keyName="variants">
      <string keyName="desc" />
      {list.map(name => <string key={name} keyName={name}/>)}
    </object>).toMatchSnapshot();
  });

  it('null', () => {
    expect(<object keyName="info">
      {null}
    </object>).toMatchSnapshot();
  });

  it('should throw error, if cannerDataType is invalid', () => {
    expect(() => <object keyName="info" type="error-type" />).toThrow('there is no type error-type');
  });
});
