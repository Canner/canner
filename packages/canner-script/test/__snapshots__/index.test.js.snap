// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`builder blog schema 1`] = `
Object {
  "connector": undefined,
  "dict": Object {},
  "graphqlClient": undefined,
  "resolvers": Object {},
  "schema": Object {
    "posts": Object {
      "description": "",
      "items": Object {
        "items": Object {
          "featureImage": Object {
            "contentType": "image",
            "description": "",
            "keyName": "featureImage",
            "packageName": "@canner/antd-object-image",
            "title": "",
            "type": "file",
            "ui": "file",
            "uiParams": Object {},
          },
          "other": Object {
            "description": "",
            "items": Object {
              "introduction": Object {
                "description": "",
                "keyName": "introduction",
                "packageName": "@canner/antd-string-input",
                "title": "",
                "type": "string",
                "ui": "input",
                "uiParams": Object {},
              },
              "position": Object {
                "description": "",
                "keyName": "position",
                "packageName": "@canner/antd-object-map",
                "title": "",
                "type": "geoPoint",
                "ui": "default",
                "uiParams": Object {},
              },
            },
            "keyName": "other",
            "packageName": "@canner/antd-object-fieldset",
            "title": "",
            "type": "object",
            "ui": "fieldset",
            "uiParams": Object {},
          },
          "pageProperty": Object {
            "description": "",
            "items": Object {
              "onTop": Object {
                "description": "",
                "keyName": "onTop",
                "packageName": "@canner/antd-boolean-switch",
                "title": "",
                "type": "boolean",
                "ui": "switch",
                "uiParams": Object {},
              },
              "order": Object {
                "description": "",
                "keyName": "order",
                "packageName": "@canner/antd-number-input",
                "title": "",
                "type": "number",
                "ui": "input",
                "uiParams": Object {},
              },
            },
            "keyName": "pageProperty",
            "packageName": "@canner/antd-object-fieldset",
            "title": "",
            "type": "object",
            "ui": "fieldset",
            "uiParams": Object {},
          },
          "share": Object {
            "description": "",
            "items": Object {
              "showGoodBottom": Object {
                "description": "",
                "keyName": "showGoodBottom",
                "packageName": "@canner/antd-boolean-switch",
                "title": "",
                "type": "boolean",
                "ui": "switch",
                "uiParams": Object {},
              },
              "showShareBottom": Object {
                "description": "",
                "keyName": "showShareBottom",
                "packageName": "@canner/antd-boolean-switch",
                "title": "",
                "type": "boolean",
                "ui": "switch",
                "uiParams": Object {},
              },
            },
            "keyName": "share",
            "packageName": "@canner/antd-object-fieldset",
            "title": "",
            "type": "object",
            "ui": "fieldset",
            "uiParams": Object {},
          },
          "status": Object {
            "description": "",
            "items": Object {
              "createAt": Object {
                "description": "",
                "keyName": "createAt",
                "packageName": "@canner/antd-string-date_time_picker",
                "title": "",
                "type": "dateTime",
                "ui": "dateTime",
                "uiParams": Object {},
              },
              "draft": Object {
                "description": "",
                "keyName": "draft",
                "packageName": "@canner/antd-boolean-switch",
                "title": "",
                "type": "boolean",
                "ui": "switch",
                "uiParams": Object {},
              },
            },
            "keyName": "status",
            "packageName": "@canner/antd-object-fieldset",
            "title": "",
            "type": "object",
            "ui": "fieldset",
            "uiParams": Object {},
          },
        },
        "type": "object",
      },
      "keyName": "posts",
      "packageName": "@canner/antd-array-table",
      "title": "",
      "toolbar": Object {
        "pagination": Object {},
      },
      "type": "array",
      "ui": "table",
      "uiParams": Object {},
    },
  },
  "storages": Object {
    "posts": Object {},
  },
  "visitors": Array [],
}
`;

exports[`builder canner.def.js function builder 1`] = `
Object {
  "builder": undefined,
  "description": "",
  "items": Object {
    "a": Object {
      "description": "",
      "keyName": "a",
      "packageName": "@canner/antd-string-input",
      "title": "",
      "type": "string",
      "ui": "input",
      "uiParams": Object {},
    },
    "d": Object {
      "description": "",
      "keyName": "d",
      "packageName": "@canner/antd-string-input",
      "title": "",
      "type": "string",
      "ui": "input",
      "uiParams": Object {},
    },
    "desc": Object {
      "description": "",
      "keyName": "desc",
      "packageName": "@canner/antd-string-input",
      "title": "",
      "type": "string",
      "ui": "input",
      "uiParams": Object {},
    },
  },
  "keyName": "variants",
  "packageName": "/Users/siou/projects/canner-script/__test__/fake-object-variants",
  "title": "",
  "type": "object",
  "ui": "fieldset",
  "uiParams": Object {},
}
`;

exports[`builder canner.def.js require builder 1`] = `
Object {
  "builder": undefined,
  "description": "",
  "items": Object {
    "a": Object {
      "description": "",
      "keyName": "a",
      "packageName": "@canner/antd-string-input",
      "title": "",
      "type": "string",
      "ui": "input",
      "uiParams": Object {},
    },
    "d": Object {
      "description": "",
      "keyName": "d",
      "packageName": "@canner/antd-string-input",
      "title": "",
      "type": "string",
      "ui": "input",
      "uiParams": Object {},
    },
    "desc": Object {
      "description": "",
      "keyName": "desc",
      "packageName": "@canner/antd-string-input",
      "title": "",
      "type": "string",
      "ui": "input",
      "uiParams": Object {},
    },
  },
  "keyName": "variants",
  "packageName": "/Users/siou/projects/canner-script/__test__/fake-object-variants",
  "title": "",
  "type": "object",
  "ui": "fieldset",
  "uiParams": Object {},
}
`;

exports[`builder function 1`] = `
Object {
  "description": "",
  "items": Object {
    "desc": Object {
      "description": "",
      "keyName": "desc",
      "packageName": "@canner/antd-string-input",
      "title": "",
      "type": "string",
      "ui": "input",
      "uiParams": Object {},
    },
    "title": Object {
      "description": "",
      "keyName": "title",
      "packageName": "@canner/antd-string-input",
      "title": "",
      "type": "string",
      "ui": "input",
      "uiParams": Object {},
    },
  },
  "keyName": "variants",
  "packageName": "@canner/antd-object-fieldset",
  "title": "",
  "type": "object",
  "ui": "fieldset",
  "uiParams": Object {},
}
`;

exports[`builder map chilren 1`] = `
Object {
  "description": "",
  "items": Object {
    "a": Object {
      "description": "",
      "keyName": "a",
      "packageName": "@canner/antd-string-input",
      "title": "",
      "type": "string",
      "ui": "input",
      "uiParams": Object {},
    },
    "b": Object {
      "description": "",
      "keyName": "b",
      "packageName": "@canner/antd-string-input",
      "title": "",
      "type": "string",
      "ui": "input",
      "uiParams": Object {},
    },
    "c": Object {
      "description": "",
      "keyName": "c",
      "packageName": "@canner/antd-string-input",
      "title": "",
      "type": "string",
      "ui": "input",
      "uiParams": Object {},
    },
    "desc": Object {
      "description": "",
      "keyName": "desc",
      "packageName": "@canner/antd-string-input",
      "title": "",
      "type": "string",
      "ui": "input",
      "uiParams": Object {},
    },
  },
  "keyName": "variants",
  "packageName": "@canner/antd-object-fieldset",
  "title": "",
  "type": "object",
  "ui": "fieldset",
  "uiParams": Object {},
}
`;

exports[`builder null 1`] = `
Object {
  "description": "",
  "keyName": "info",
  "packageName": "@canner/antd-object-fieldset",
  "title": "",
  "type": "object",
  "ui": "fieldset",
  "uiParams": Object {},
}
`;
